[SelectionDAG] fold (not (sub Y, X)) -> (add X, ~Y)

AZero13 opened the pull request:
This replaces (not (neg x)) -> (add X, -1) because that is covered by this.

badumbatish commented:
has there been a test for this selectiondag folding after & before the TODO? it would be beneficial to add one if not

topperc commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -9979,13 +9979,16 @@ SDValue DAGCombiner::visitXOR(SDNode *N) {
-      isNullConstant(N0.getOperand(0))) {
-    return DAG.getNode(ISD::ADD, DL, VT, N0.getOperand(1),
-                       DAG.getAllOnesConstant(DL, VT));
+  // fold (not (sub Y, X)) -> (add X, ~Y) is subtract is a single use
```
```suggestion
  // fold (not (sub Y, X)) -> (add X, ~Y) if subtract is a single use
```

topperc commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -9979,13 +9979,16 @@ SDValue DAGCombiner::visitXOR(SDNode *N) {
-      isNullConstant(N0.getOperand(0))) {
-    return DAG.getNode(ISD::ADD, DL, VT, N0.getOperand(1),
-                       DAG.getAllOnesConstant(DL, VT));
+  // fold (not (sub Y, X)) -> (add X, ~Y) if subtract is a single use
```
```suggestion
  // fold (not (sub Y, X)) -> (add X, ~Y) if subtract has a single use
```

AZero13 committed:
[SelectionDAG] fold (not (sub Y, X)) -> (add X, ~Y)

This replaces (not (neg x)) -> (add X, -1) because that is covered by this.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -9979,13 +9979,15 @@ SDValue DAGCombiner::visitXOR(SDNode *N) {
+  // fold (not (sub Y, X)) -> (add X, ~Y) if Y is a constant.
+  // FIXME: We can also do this with single-use sub, but this causes an infinite
+  // loop
+  if (isAllOnesConstant(N1) && N0.getOpcode() == ISD::SUB) {
```
```suggestion
  if (N0.getOpcode() == ISD::SUB && isAllOnesConstant(N1)) {
```

arsenm reviewed (changes requested):
Missing tests. Please never post functional change patches without tests

topperc commented:
Stop force pushing patches. I'm tired of repeating this.

AZero13 committed:
Allow fold to be on platforms that do not have and-not

Otherwise we will end up in an infinite loop. Yes, this should in theory not be whenever there is or-not, and eor-not, or whatever, but in practice this is only an issue for aarch64, and that simply has the same rules as and-not.

AZero13 committed:
Fix comment and test

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -9980,12 +9980,19 @@ SDValue DAGCombiner::visitXOR(SDNode *N) {
-      SDValue NotY = DAG.getNOT(DL, N00, VT); // N00 = ~N00
+    if (isa<ConstantSDNode>(N00) || (N00.hasOneUse() && !TLI.hasAndNot(N01))) {
+      SDValue NotY =
+          DAG.getNode(ISD::XOR, SDLoc(N00), VT, N00, N1); // N00 = ~N00
```
Still better to use getNOT for clarity
